builder/resubmit_utils.go

```diff
@@ -9,13 +9,28 @@ import (
 )
 
 // runResubmitLoop checks for update signal and calls submit respecting provided rate limiter and context
-func runResubmitLoop(ctx context.Context, limiter *rate.Limiter, updateSignal chan struct{}, submit func()) {
+func runResubmitLoop(ctx context.Context, limiter *rate.Limiter, updateSignal chan struct{}, submit func(), submitTime *time.Time) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-updateSignal:
-			res := limiter.Reserve()
+			var res *rate.Reservation
+			if submitTime == nil {
+				res = limiter.Reserve()
+			} else {
+				now := time.Now()
+				sleepTime := submitTime.UTC().UnixMilli() - now.UTC().UnixMilli()
+				log.Debug("resubmit loop",
+					"now-sec", now.UTC().Unix(),
+					"slot", ctx.Value(key("slot")),
+					"slot-sec", submitTime.Add(4*time.Second).Unix(),
+					"delta-now-from-slot-ms", submitTime.Add(4*time.Second).Sub(now).Milliseconds(),
+					"block-time-sec", ctx.Value(key("timestamp")))
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
```

Comment:
Small nit: sleep could be interrupted by context cancellation, maybe it makes sense to select on `time.After` and `context.Done()`

Reply:
good catch, updated